pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,656 @@
+	if !h.localInformers.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	lcluster, err := request.ValidClusterFrom(ctx)
+	if err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	if sets.NewString(user.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+		// If we are the system privileged group, it might be a call from the virtual workspace
+		// in which case we also search the user in the soft impersonation header.
+		if impersonated, err := softimpersonation.UserInfoFromRequestHeader(req); err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		} else if impersonated != nil {
+			user = impersonated
+		}
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
+		return
+	}
+
+	if lcluster.Name == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+
+		// if we are in the special case of a `kubectl get workspace ~` request (against the 'root' workspace),
+		// we return the Home workspace definition of the current user, possibly even before its
+		// underlying ClusterWorkspace resource exists.
+
+		homeLogicalClusterName := h.getHomeLogicalClusterName(user.GetName())
+		homeClusterWorkspace, err := h.localInformers.getClusterWorkspace(homeLogicalClusterName)
+		if err != nil && !kerrors.IsNotFound(err) {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he requests to see its definition.
+			// We don't need to check any permission before returning the home workspace definition since,
+			// once it has been created, a home workspace is owned by the user.
+
+			homeWorkspace := &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+			responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+			return
+		}
+
+		// Test if the user has the right to see his Home workspace even though it doesn't exists
+		// => test the get verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+		attributes := homeWorkspaceAuthorizerAttributes(user, "get")
+		decision, reason, err := h.authz.Authorize(ctx, attributes)
+		if err != nil {
+			klog.Errorf("failed to authorize user %q to get a home workspace %q: %w", user.GetName(), lcluster.Name, err)
```

Comment:
unconditional klogs in the handler chain are always questionable. Better call `runtime.HandleError` as that throttles the output.

Reply:
done